src/lang/processor.rs

```diff
@@ -0,0 +1,1181 @@
+            request_pair: None,
+        }
+    }
+}
+
+pub fn process_pest_tree(
+    mut pairs_iterator_for_document_rule: Pairs<'_, Rule>,
+    params: &Params,
+) -> Result<ProcessedOutput, ProcessorError> {
+    let mut processing_ctx = ProcessingContext::new(params);
+
+    let document_node = pairs_iterator_for_document_rule.next().ok_or_else(|| {
+        ProcessorError::Internal(format!(
+            "Parser returned no pairs for the expected top-level rule: {:?}.",
+            Rule::document
+        ))
+    })?;
+
+    if document_node.as_rule() != Rule::document {
+        return Err(ProcessorError::Internal(format!(
+            "Expected top-level pair to be Rule::{:?}, but found Rule::{:?}.",
+            Rule::document,
+            document_node.as_rule()
+        )));
+    }
+
+    let document_content_pairs = document_node.into_inner();
+
+    first_pass(document_content_pairs, &mut processing_ctx)?;
+
+    second_pass(&mut processing_ctx)
+}
+
+/// Pass 1: Iterates through top-level definitions, records custom predicate
+/// signatures and stores pairs for Pass 2.
+fn first_pass<'a>(
+    document_pairs: Pairs<'a, Rule>,
+    ctx: &mut ProcessingContext<'a>,
+) -> Result<(), ProcessorError> {
+    let mut defined_custom_names: HashSet<String> = HashSet::new();
+    let mut first_request_span: Option<(usize, usize)> = None;
+
+    for pair in document_pairs {
+        match pair.as_rule() {
+            Rule::custom_predicate_def => {
+                let pred_name_pair = pair
+                    .clone()
+                    .into_inner()
+                    .find(|p| p.as_rule() == Rule::identifier)
+                    .unwrap();
+                let pred_name = pred_name_pair.as_str().to_string();
+
+                if defined_custom_names.contains(&pred_name) {
+                    return Err(ProcessorError::DuplicateDefinition {
+                        name: pred_name,
+                        span: Some(get_span(&pred_name_pair)),
+                    });
+                }
+                defined_custom_names.insert(pred_name.clone());
+
+                let public_arity = count_public_args(&pair)?;
+                ctx.custom_predicate_signatures.insert(
+                    pred_name.clone(),
+                    (ctx.custom_predicate_pairs.len(), public_arity),
+                );
+                ctx.custom_predicate_pairs.push(pair);
+            }
+            Rule::request_def => {
+                if ctx.request_pair.is_some() {
+                    return Err(ProcessorError::MultipleRequestDefinitions {
+                        first_span: first_request_span,
+                        second_span: Some(get_span(&pair)),
+                    });
+                }
+                first_request_span = Some(get_span(&pair));
+                ctx.request_pair = Some(pair);
+            }
+            Rule::EOI => break,
+            Rule::COMMENT | Rule::WHITESPACE => {}
+            _ => {
```

Comment:
We call `first_pass` after checking we're in a `document`, and by definition there are no more rules in `document`, so this should be `unreachable` instead of error